File: Sujin/Ch13_Graph&Tree&Dijkstra/543_Diameter_of_Binary_Tree.cpp
#include <iostream>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int ans = 0; // answer

    int DFS(TreeNode* node){
        if(node==NULL) return 0;

        int left = DFS(node->left);
        int right = DFS(node->right);

        ans = max(ans, left+right); // 최댓값 갱신
        return max(left, right)+1; // DFS내에서 반환해야하는 값
    }

    int diameterOfBinaryTree(TreeNode* root){
        if(root==NULL) return 0; // root==[]인 경우의 예외처리

        DFS(root);
        return ans;
    }
};

File: Suyeon/ch14_Tree/543_Diameter_of_Binary_Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    result = 0
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:

        def maxheight(root):
            if root is None:
                return 0
            left = maxheight(root.left)
            right = maxheight(root.right)
            self.result = max(Solution.result, left + right)
            return max(left, right) + 1

        maxheight(root)

        return self.result # 중첩함수에서 자식함수가 부모함수의 값을 재할할 수 없기 때문에 클래스변수로 선언